gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -272,24 +350,108 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
         throw ReflectionHelper.createExceptionForUnexpectedIllegalAccess(e);
       }
       in.endObject();
-      return instance;
+      return finalize(intermediary);
     }
 
-    @Override public void write(JsonWriter out, T value) throws IOException {
-      if (value == null) {
-        out.nullValue();
-        return;
+    /** Create the Object that will be used to collect each field value */
+    abstract I createIntermediary();
+    /**
+     * Read a single Bounded field into the intermediary. The JsonReader will be pointed at the
+     * start of the value for the BoundField to read from.
+     */
+    abstract void readField(I intermediary, JsonReader in, BoundField field)
+        throws IllegalAccessException, IOException;
+    /** Convert the intermediary to a final instance of T. */
+    abstract T finalize(I intermediary);
+  }
+
+  private static final class FieldReflectionAdapter<T> extends Adapter<T, T> {
+    private final ObjectConstructor<T> constructor;
+
+    FieldReflectionAdapter(ObjectConstructor<T> constructor, Map<String, BoundField> boundFields) {
+      super(boundFields);
+      this.constructor = constructor;
+    }
+
+    @Override
+    T createIntermediary() {
+      return constructor.construct();
+    }
+
+    @Override
+    void readField(T intermediary, JsonReader in, BoundField field)
+        throws IllegalAccessException, IOException {
+      field.readIntoField(in, intermediary);
+    }
+
+    @Override
+    T finalize(T intermediary) {
+      return intermediary;
+    }
+  }
+
+  private static final class RecordAdapter<T> extends Adapter<T, Object[]> {
+    // The actual record constructor.
+    private final Constructor<? super T> constructor;
+    // Array of arguments to the constructor, initialized with default values for primitives
+    private final Object[] constructorArgsDefaults;
+    // Map from field names to index into the constructors arguments.
+    private final Map<String, Integer> fieldIndices = new HashMap<>();
```

Comment:
Rather minor, but maybe here and also below it would be better to use the name "component index". Fields (even for records) don't have any guaranteed order or index, but record components do have a specified order.

Reply:
Agree.